proxy/src/telemetry/metrics/prometheus.rs

```diff
@@ -0,0 +1,469 @@
+                self.update(|metrics| {
+                    *metrics.response_total(&labels) += 1;
+                    *metrics.response_duration(&labels) += fail.since_response_open;
+                    *metrics.response_latency(&labels) += fail.since_request_open;
+                });
+            },
+
+            Event::TransportOpen(_) | Event::TransportClose(_, _) => {
+                // TODO: we don't collect any metrics around transport events.
+            },
+        };
+    }
+}
+
+
+// ===== impl Serve =====
+
+impl Serve {
+    fn new(metrics: &Arc<Mutex<Metrics>>) -> Self {
+        Serve { metrics: metrics.clone() }
+    }
+}
+
+impl HyperService for Serve {
+    type Request = HyperRequest;
+    type Response = HyperResponse;
+    type Error = hyper::Error;
+    type Future = FutureResult<Self::Response, Self::Error>;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+        if req.path() != "/metrics" {
+            return future::ok(HyperResponse::new()
+                .with_status(StatusCode::NotFound));
+        }
+
+        let body = {
+            let metrics = self.metrics.lock()
+                .expect("metrics lock poisoned");
+            format!("{}", *metrics)
+        };
+        future::ok(HyperResponse::new()
+            .with_header(ContentLength(body.len() as u64))
+            .with_header(ContentType::plaintext())
+            .with_body(body))
+    }
+}
+
+
+// ===== impl RequestLabels =====
+
+impl<'a> RequestLabels {
+    fn new(req: &ctx::http::Request) -> Self {
+        let outbound_labels = if req.server.proxy.is_outbound() {
+            Some(OutboundLabels {
+                // TODO: when #429 is done, add appropriate destination label.
+                ..Default::default()
+            })
+        } else {
+            None
+        };
+
+        let authority = req.uri
+            .authority_part()
+            .map(http::uri::Authority::to_string)
+            .unwrap_or_else(String::new);
+
+        RequestLabels {
+            outbound_labels,
+            authority,
+            ..Default::default()
+        }
+    }
+}
+
+impl fmt::Display for RequestLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "authority=\"{}\",", self.authority)?;
+        if let Some(ref outbound) = self.outbound_labels {
+            write!(f, "direction=\"outbound\",{}", outbound)?;
+        } else {
+            write!(f, "direction=\"inbound\"")?;
+        }
+        // TODO: when PR #448 lands, use CONDUIT_PROMETHEUS_LABELS
+        // to add the appropriate labels for the owner of this pod.
+
+        Ok(())
+    }
+
+}
+
+impl fmt::Display for OutboundLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if let Some(ref ns) = self.namespace {
+            write!(f, "dst_namespace=\"{}\"", ns)?;
+        }
+        match *self {
+            OutboundLabels { deployment: Some(ref d), .. } =>
```

Comment:
Does this mean that the `OutboundLabels` can only ever have 1 of these values set to `Some`?

Reply:
According to @adleong, pod owners are mutually exclusive, so yes. `OutboundLabels` should probably be an `enum` to better encode that, I left it like this because currently it's not set at all and I wanted everything to work regardless.

Reply:
If they are mutually exclusive, then an `enum` conveys that much more clearly. When reading this match, I wasn't sure if they were supposed to be exclusive, or if this was a bug that somehow tried to handle all combinations of `Some`s.

Reply:
Yeah, I agree that it should be an `enum`; I was planning on making that change once the information was available to actually set a value for `OutboundLabels`.

Reply:
The only thing is that it's possible for Future Us to forget, and possibly not adjust this, and then Really Future Us changes the contract of `OutboundLabels` and this is then broken. (I've been that Future Me before :cry:)

It seems safer to either 1) write this as an `enum` now, or 2) clear out `OutboundLabels` of all fields, since we may not know what it actually looks like yet.

Reply:
That's fair. How about an `enum` with a `None` variant or something, which can be removed later once we have the information to actually set a value?

Reply:
That seems fine!

Reply:
Just made the field optional for now, instead. What do you think of 6e454d52a1407d1203a83faa58142486af9d4347?